ports/stm32f4/supervisor/internal_flash.c

```diff
@@ -50,21 +97,105 @@ uint32_t supervisor_flash_get_block_size(void) {
 }
 
 uint32_t supervisor_flash_get_block_count(void) {
-    return false;
+    return INTERNAL_FLASH_FILESYSTEM_NUM_BLOCKS;
 }
 
 void supervisor_flash_flush(void) {
+}
 
+static int32_t convert_block_to_flash_addr(uint32_t block) {
+    if (0 <= block && block < INTERNAL_FLASH_FILESYSTEM_NUM_BLOCKS) {
+        // a block in partition 1
+        return INTERNAL_FLASH_FILESYSTEM_START_ADDR + block * FILESYSTEM_BLOCK_SIZE;
+    }
+    // bad block
+    return -1;
 }
 
 mp_uint_t supervisor_flash_read_blocks(uint8_t *dest, uint32_t block, uint32_t num_blocks) {
-
+	int32_t src = convert_block_to_flash_addr(block);
+	if (src == -1) {
+        // bad block number
+        return false;
+    }
+    memcpy(dest, (uint8_t*) src, FILESYSTEM_BLOCK_SIZE*num_blocks);
     return 0; // success
 }
 
-mp_uint_t supervisor_flash_write_blocks(const uint8_t *src, uint32_t lba, uint32_t num_blocks) {
+bool supervisor_flash_write_block(const uint8_t *src, uint32_t block) {
+    int32_t dest = convert_block_to_flash_addr(block);
+    if (dest == -1) {
+        // bad block number
+        mp_printf(&mp_plat_print, "BAD FLASH BLOCK ERROR");
+        return false;
+    }
+
+    // unlock flash
+    HAL_FLASH_Unlock();
+
+    // set up for erase
+    FLASH_EraseInitTypeDef EraseInitStruct;
+    EraseInitStruct.TypeErase = TYPEERASE_SECTORS;
+    EraseInitStruct.VoltageRange = VOLTAGE_RANGE_3; // voltage range needs to be 2.7V to 3.6V
+    // get the sector information
+    uint32_t sector_size;
+    uint32_t sector_start_addr;
+    EraseInitStruct.Sector = flash_get_sector_info(dest, &sector_start_addr, &sector_size);
+    EraseInitStruct.NbSectors = 1;
+    if (sector_size>0x4000) return false;
+
+    // copy the sector
+    memcpy(sector_copy,(void *)sector_start_addr,sector_size);
+
+    // // overwrite sector data
+    memcpy(sector_copy+(dest-sector_start_addr),src,FILESYSTEM_BLOCK_SIZE);
+
+    // find end address, subtract for number of sectors
+    // Shouldn't be required since blocks will always fit in a single sector, they should never overlap
+    //EraseInitStruct.NbSectors = flash_get_sector_info(dest + FILESYSTEM_BLOCK_SIZE - 1, NULL, NULL) - EraseInitStruct.Sector + 1;
+
+    // erase the sector
+    uint32_t SectorError = 0;
+    if (HAL_FLASHEx_Erase(&EraseInitStruct, &SectorError) != HAL_OK) {
+        // error occurred during sector erase
+        HAL_FLASH_Lock(); // lock the flash
+        mp_printf(&mp_plat_print, "FLASH SECTOR ERASE ERROR");
+        return false;
+    }
+
+	__HAL_FLASH_DATA_CACHE_DISABLE();
```

Comment:
Indentation. Use spaces instead of tabs. Looks like you're using 4-space tabs?

Reply:
Yes, looks like some of these snuck in when copying the occasional segment from the HAL docs or CubeMX setup.